In [1]:
import pandas as pd
import numpy as np
from sklearn import tree
from sklearn import preprocessing

C:\Users\OMKAR\anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\OMKAR\anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\OMKAR\anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)
C:\Users\OMKAR\anaconda3\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject
  return f(*args, **kwds)


# importing the data


In [2]:
data=pd.read_excel("Bank_Personal_Loan_Modelling.xlsx",sheet_name=1)
data.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

# preprosessing the data

In [3]:
data.shape

(5000, 14)

In [4]:
data.isna().sum()

ID                    0
Age                   0
Experience            0
Income                0
ZIP Code              0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal Loan         0
Securities Account    0
CD Account            0
Online                0
CreditCard            0
dtype: int64

In [5]:
data.dtypes

ID                      int64
Age                     int64
Experience              int64
Income                  int64
ZIP Code                int64
Family                  int64
CCAvg                 float64
Education               int64
Mortgage                int64
Personal Loan           int64
Securities Account      int64
CD Account              int64
Online                  int64
CreditCard              int64
dtype: object

<h2>Imorting Random forest algorithum</h2>

    For insampling of decision tree to find the important variable from dataset


In [10]:
from sklearn.ensemble import RandomForestClassifier

rfModel=RandomForestClassifier(oob_score=True)



<h3>Defining Featires for Random forest</h3>

In [11]:
data.columns

Index(['ID', 'Age', 'Experience', 'Income', 'ZIP Code', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Personal Loan', 'Securities Account',
       'CD Account', 'Online', 'CreditCard'],
      dtype='object')

In [12]:
features=['ID', 'Age', 'Experience', 'Income', 'ZIP Code', 'Family', 'CCAvg',
       'Education', 'Mortgage','Securities Account',
       'CD Account', 'Online', 'CreditCard']

rfModel.fit(X=data[features],y=data["Personal Loan"]);

In [14]:
print("Oob score for this model is  :");
print(rfModel.oob_score_)

Oob score for this model is  :
0.986


In [15]:
for feature,imp in zip(features,rfModel.feature_importances_):
    print(feature,"     :",imp)

ID      : 0.03231451663215451
Age      : 0.031197636509472333
Experience      : 0.02935997838913724
Income      : 0.3329364240989184
ZIP Code      : 0.031988758935884605
Family      : 0.10318660034284212
CCAvg      : 0.15456390686409077
Education      : 0.18816395348817672
Mortgage      : 0.02880748599407905
Securities Account      : 0.004001597878736942
CD Account      : 0.050879900438484935
Online      : 0.006051719193737926
CreditCard      : 0.0065475212342844605


# From the RF-Model we can see that factors effecting to personal Loan is :
Income : 33%

CCAcg,Education : more than 15%

<h3> We can build Decision tree Model on this features</h3>

In [18]:
# For decision tree importing required libraries  :


from sklearn import tree


treeModel = tree.DecisionTreeClassifier()

impPredictors=pd.DataFrame([data["Income"],data["CCAvg"],data["Education"]]).T;

treeModel.fit(X=impPredictors,y=data["Personal Loan"]);

In [19]:
with open('Dtree.dot','w') as f:  #decision tree always store in .dot format
    f=tree.export_graphviz(treeModel,feature_names=["Income","CCAverage","Education"],out_file=f);


In [21]:
treeModel.score(X=impPredictors,y=data["Personal Loan"])


0.9972

# conclusion


The Model that we build is working fine as it gives the accuracy of 99.72%